### PR TITLE
Upgrade to Python 3.9

### LINKED ISSUE #1181: Upgrade to Python 3.9
As a developer, I'd like to use the latest stable version of python so that we can upgrade our build environment

Acceptance criteria
- [ ] Update runtime.txt to update the heroku stack
- [ ] Update the github actions configuration to use this version of python
- [ ] Update docker containers to use this version of python

### PR DESCRIPTION
**Pull request by umarmughal824**
#### What are the relevant tickets?
fixes #1181 

#### What's this PR do?
Upgrade to Python 3.9

#### How should this be manually tested?

- [ ] Build the docker containers
- [ ] project should still be functional
- [ ] Python version should point to 3.9.4

#### Where should the reviewer start?
(Optional)

#### Any background context you want to provide?
(Optional)

#### Screenshots (if appropriate)
(Optional)

#### What GIF best describes this PR or how it makes you feel?
(Optional)

### CODE REVIEW

**Review by arslanashraf7 — CHANGES_REQUESTED** [review]
@umarmughal824, There are a few things that might still need updating/upgrading in this PR.

1) Compiling the requirements still generates updated `requirements.txt` and removes some packages, so it might be good to do that in this PR.
2) We faced an issue with `newrelic` after `Python 3.9` upgrade and we had to upgrade its version too (Ref: https://github.com/mitodl/micromasters/pull/4799)
3) I think we'll also need to update `iPython` since the shell throws an error, try `./mange.py shell` in your container.
4) We should upgrade https://github.com/mitodl/bootcamp-ecommerce/blob/umar/bc-1181/pyproject.toml#L2 to support python 3.9.

**Comment by umarmughal824** [comment]
@arslanashraf7 I'm getting that error while rebasing to the update master after merging your PR so that's why could not resolve merge conflicts.
```
ERROR: Could not find a version that satisfies the requirement djangorestframework-simplejwt==4.4.0 (from versions: 1.0, 1.1, 1.2, 1.2.1, 1.3, 1.4, 1.5.1, 2.0.0, 2.0.1, 2.0.2, 2.0.3, 2.0.4, 2.0.5, 2.1, 3.0, 3.1, 3.2, 3.2.1, 3.2.2, 3.2.3, 3.3, 4.0.0, 4.1.0, 4.1.1, 4.1.2, 4.1.3, 4.1.4, 4.1.5, 4.2.0, 4.3.0, 4.5.0, 4.6.0)
ERROR: No matching distribution found for djangorestframework-simplejwt==4.4.0
```

**Comment by arslanashraf7** [comment]
> ```
> ERROR: No matching distribution found for djangorestframework-simplejwt==4.4.0
> ```

@umarmughal824 The issue seems to be related to `Python 3.9`. Can you please have a look at https://github.com/jazzband/django-rest-framework-simplejwt/issues/315. It looks like we'll need to update the version for `djangorestframework-simplejwt`.

**Comment by umarmughal824** [comment]
> > ```
> > ERROR: No matching distribution found for djangorestframework-simplejwt==4.4.0
> > ```
> 
> @umarmughal824 The issue seems to be related to `Python 3.9`. Can you please have a look at [jazzband/django-rest-framework-simplejwt#315](https://github.com/jazzband/django-rest-framework-simplejwt/issues/315). It looks like we'll need to update the version for `djangorestframework-simplejwt`.

Sure

**Review by arslanashraf7 — COMMENTED** [review]
@umarmughal824 I was testing the things with PR today and the issue with the shell still exists which is working in the prior release before the python 3.9 upgrades. My wild guess is that it is related to `iPython`. Screenshot attached.

Screenshot 2021-04-28 at 5 33 46 PM

**Comment by umarmughal824** [comment]
@arslanashraf7 thanks for pointing it out. It should be functional now
Screenshot 2021-04-28 at 18 22 03

**Review by arslanashraf7 — APPROVED** [review]
Looks good 👍 , Thanks for making the changes.

File: hubspot/api.py [hunk 1]
```diff
@@ -182,7 +182,10 @@ def get_sync_errors(limit=200, offset=0):
     while len(errors) > 0:
         yield from errors
         offset += limit
-        errors = paged_sync_errors(limit, offset)
+        try:
+            errors = paged_sync_errors(limit, offset)
+        except StopIteration:
```
Comment (mbertrand) [inline hubspot/api.py:187]: Does `paged_sync_errors` actually throw a `StopIteration` here sometimes? I'd expect it to maybe throw an `HTTPError`.

Comment (umarmughal824) [reply hubspot/api.py:187]: it was throwing StopIteration and test cases were failing because of it that's why I have added that.

Comment (mbertrand) [reply hubspot/api.py:187]: 👍

**Review by mbertrand — APPROVED** [review]
👍

**Pull request merged**